File: 21MergeTwoSortedLists/21MergeTwoSortedLists/21MergeTwoSortedLists.cpp
#include <iostream>

using namespace std;


 // Definition for singly-linked list.
 struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
 };
 
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        int val1, val2;
        ListNode* list3 = new ListNode();
        ListNode* current1 = list1; // head of list1
        ListNode* current2 = list2; // head of list2
        ListNode* current3 = list3;
        if (current1 != nullptr || current2 != nullptr)
            while (current1->next != NULL || current2 != NULL) {
                //current1 = current1->next;
                //current2 = current2->next;
                if (current1->next != nullptr) {
                    current1 = current1->next;
                    val1 = current1->val;
                }
                else {
                    val1 = 101;
                }
                if (current2->next != nullptr) {
                    current2 = current2->next;
                    val2 = current2->val;
                }
                else {
                    val2 = 101;
                }

                if (val2 >= val1) {
                    ListNode* newNode = new struct ListNode(val1);
                    current3->next = newNode;
                }
                else {
                    ListNode* newNode = new struct ListNode(val2);
                    current3->next = newNode;
                }
            }
        return list3;
    }
};

int main() {
    Solution sol = new Solution();

}

File: LinkedList/21MergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        h1 = list1
        h2 = list2
        list3 = None

        # base case
        if(h1 != None and h2!= None):
            if(h1.val < h2.val):
                list3 = ListNode(h1.val, next=None)
                h1 = h1.next
            else:
                list3 = ListNode(h2.val, next=None)
                h2 = h2.next
        elif h1 != None:
            list3 = ListNode(h1.val, next=None)
            h1 = h1.next
        elif h2 != None:
            list3 = ListNode(h2.val, next=None)
            h2 = h2.next
        else:
            return list3

        # always use a temp pointer to iterate through a linked list, not the list variable itself
        curr = list3
        while(h1 != None or h2 != None):
            if(h1 != None and h2 != None and h1.val < h2.val):
                curr.next = ListNode(h1.val, next=None)
                h1 = h1.next
            elif(h1 != None and h2 != None and h1.val >= h2.val):
                curr.next = ListNode(h2.val, next=None)
                h2 = h2.next
            elif(h1 != None):
                curr.next = ListNode(h1.val, next=None)
                h1 = h1.next
            elif(h2 != None):
                curr.next = ListNode(h2.val, next=None)
                h2 = h2.next
            
            curr = curr.next
        
        return list3